File: java/Basics/Variables.java
package Basics;

public class Variables {
    public static void main(String [] args) {
        /*
            1. To create a variable, you must specify the type and assign it a value:
                [type] variableName = value;
            For example, create a variable name of type String and assign it the value: "Jeming"
        */
        String name = "Jeming";
        System.out.println(name);
        /*
            2. To create a variable that stores an integer:
        */
        int number01 = 5;
        Integer number02 = 10;
        System.out.println(number01);
        System.out.println(number02);
        System.out.println("Sum of: " + number01 + " and " + number02 + " is: " + (number01 + number02));
        /*
            3. You can also declare a variable without assigning the value, and assign the value later:
        */
        int number;
        number = 20;
        System.out.println(number);
        /*
            4. You can also assign a new value to an existing variable, overwriting the previous value:
        */
        int numberAgain = 3;
        numberAgain = 999;
        System.out.println(numberAgain);
        /*
            5. If you don't want a variable to be overwrited again, use the "final" keyword,
            which will make the variable unchangeable and read-only
        */
        final int finalNumber = 33;
        System.out.println("Final Number: " + finalNumber);
        /*
            6. Other Variable Types:
        */
        // Numeric Types
        short shortType = 2;        // Size of 2 Bytes
        int integerType = 4;        // Size of 4 Bytes
        long longType = 8;          // Size of 8 Bytes
        float floatType = 3.00f;    // Size of 4 Bytes - Stores fractional numbers. Value should end with an "f" for floats and "d" for doubles

        double doubleType = 8.9231; // Size of 8 Bytes - Stores fractional numbers    
        // Characters/String Types
        char charType = 'C';        // Must be in single quotes ''
        String stringType = "String";
        // Boolean Type
        boolean booleanType = true;
        /*
            7. Declaring many variables in ONE LINE
        */
        int x = 5; int y = 10; int z = 15;
        System.out.println(x + y + z);
        /*
            8. Type Casting. There are two ways of type casting:
                i) Widening Casting (automatically) - converting a smaller type to a larger type size
                    byte -> short -> char -> int -> long -> float -> double
                ii) Narrowing Casting (manually) - converting a larger type to a smaller size type
                    double -> float -> long -> int -> char -> short -> byte
        */
        // Widening Casting - done automatically when passing a smaller size type to a larger size type
        int cast01 = 9;
        double double01 = cast01;
        System.out.println(cast01);
        System.out.println(double01);
        // Narrowing Casting - done manully by placing the type in parentheses in front of the value:
        int cast02 = 99;
        double double02 = (double) cast02;
        System.out.println(cast02);
        System.out.println(double02);
    }
}

File: kotlin/src/main/kotlin/Basics/Variables.kt
package Basics

/* ==================== Variables ==================== */

/*
    The main difference between Java and Kotlin is that type in Kotlin are NON-NULL by default.
    This means there is a distinct difference between, for example, a String, and a Nullable-String.

    https://stackoverflow.com/questions/52964417/how-to-create-variable-in-kotlin-without-initialization

    In Java, when you declare a variable but don't initialise it (i.e. you don't assign a value to it),
    the variable will implicitly be declared as "null". Kotlin doesn't automatically have this
    behaviour due to the variables being non-null by default.
 */
/*
    For example, the variable below called "cheeseString" is a Non-Null String.
    This means if we try to assign this variable to equal "null" (shown in the line that has been
    commented out below), you'll get an error:
*/
//var cheeseString: String = null
/*
    In order to declare a variable in Kotlin and not initialise/assign it a value, you must use a
    "?" at the end of the variable type, and set it equal to null:
*/
var cheesyString: String? = null

/*
    Variables can also be declared outside any type of enclosing FUNCTION or CLASS.
    Variables like this are known as "Top-Level Variables"
*/
val topLevelVariable: String = "I'm a 'Top-Level' String Variable :)"

fun main() {
    println(topLevelVariable)

    /*
        There are two keywords for assigning variables in Kotlin:
            1. val - read-only local variables which can only be assigned a value ONCE. They cannot
                     be re-assigned to another value.
            2. var - variables which can be re-assigned

        The general syntax to assign a NON-NULL variable in Kotlin is:
            val [variableName]: [variableType] = [value]
            var [variableName]: [variableType] = [value]

        The general syntax to assign a NULLABLE variable (i.e. a variable which can be equal to
        null when you declare it, or when you re-assign it) in Kotlin is:
            val [variableName]: [variableType]? = null
            var [variableName]: [variableType]? = null
    */

    val sentence0: String = "This is a sentence"
    println(sentence0)

    var sentence1: String = "A quick brown fox"
    // re-assigning the "var" variable
    sentence1 = "jumped over the lazy dog"
    println(sentence1)

    /*
        Kotlin can also INFER type declarations. So you are not required to specify the variable
        type when you declare a variable.

        The syntax to assign variables where the type will be inferred is:
            val [variableName] = [value]

        However, if you want to declare a nullable variable (i.e. a variable you want to assign to
        be as null when you declare it, or when you re-assign it), you must specify the type.
    */
    val sentence2 = "Good evening"

    /*
        Variable Types in Kotlin are the same as Java, but with the names all starting with capitals
            • String
            • Char
            • Int
            • Float
            • Double
            • Boolean
    */
    var sentenceVariable: String = "A sentence in Kotlin"
    var characterVariable: Char = 'Z'
    var integerVariable: Int = 23
    var floatVariable: Float = 63.32f
    var doubleVariable: Double = 84.32
    var booleanVariable: Boolean = false

}